File: ts/src/adapters/goose.ts
import { register } from '../registry.js'
import { HarnessError, finalizeCommand, validateRunSpec } from '../base.js'
import type { Adapter, BuildCommand, ParsedOutput, RunSpec, SubprocOutcome } from '../base.js'

/** Child env that makes goose auto-approve tool calls; `approve` prompts and rejects headlessly. */
const BYPASS_MODE = 'auto'

type StreamEvent = Record<string, unknown> & { type: string }

function isStreamEvent(value: unknown): value is StreamEvent {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
    && typeof (value as Record<string, unknown>)['type'] === 'string'
}


/** Goose's local non-interactive run command, not its ACP/HTTP server. */
const gooseAdapter: Adapter = {
  name: 'goose',
  // Use --system rather than projecting a file that CONTEXT_FILE_NAMES could exclude.
  instructionsFilename: '',
  defaultModel: '', // Provider and model come from the caller's goose config / GOOSE_* env.
  // Bypass is env, not argv: GOOSE_MODE=auto on the child.
  permissionBypassArgs: [],
  configHomeEnv: 'GOOSE_PATH_ROOT',

  buildCommand(spec: RunSpec): BuildCommand {
    const validated = validateRunSpec(this, spec)
    const env: Record<string, string> = {}
    if (validated.permissionPolicy === 'bypass') {
      const explicit = spec.env?.['GOOSE_MODE']
      if (explicit !== undefined && explicit !== BYPASS_MODE) {
        throw new HarnessError(
          `permissionPolicy "bypass" conflicts with env.GOOSE_MODE=${JSON.stringify(explicit)}; set one`,
          'invalid-options',
        )
      }
      env['GOOSE_MODE'] = BYPASS_MODE
    }
    const args = ['run', '--quiet', '--output-format', 'stream-json']
    if (validated.model) args.push('--model', validated.model)
    if (spec.instructions != null) args.push(`--system=${spec.instructions}`)
    args.push(`--text=${spec.prompt}`)
    return finalizeCommand(this, spec, validated, { cmd: 'goose', args, env, model: spec.model || null })
  },

  parseOutput(_spec: RunSpec, outcome: SubprocOutcome): ParsedOutput {
    const events: StreamEvent[] = []
    let complete: StreamEvent | null = null
    for (const line of outcome.stdout.split('\n')) {
      const trimmed = line.trim()
      if (!trimmed.startsWith('{')) continue
      let event: unknown
      try {
        event = JSON.parse(trimmed)
      } catch {
        continue
      }
      if (!isStreamEvent(event)) continue
      events.push(event)
      if (event.type === 'complete') complete = event
    }
    // The last `complete` carries cumulative session usage; never sum events.
    // Each metric stands alone: typeof number excludes booleans, the range check excludes NaN/Infinity.
    const cost = complete?.['cost_usd']
    const tokensIn = complete?.['input_tokens']
    const tokensOut = complete?.['output_tokens']
    return {
      costUsd: typeof cost === 'number' && Number.isFinite(cost) && cost >= 0 ? cost : null,
      tokensIn: typeof tokensIn === 'number' && Number.isInteger(tokensIn) && tokensIn >= 0 ? tokensIn : null,
      tokensOut: typeof tokensOut === 'number' && Number.isInteger(tokensOut) && tokensOut >= 0 ? tokensOut : null,
      raw: events.length === 0 ? null : events,
    }
  },
}

register('goose', gooseAdapter)

File: src/harness/adapters/goose.py
"""Goose headless JSONL. The last `complete` contains cumulative session usage.

Non-JSON banners and partial lines are ignored; events remain in `raw`, including
provider errors that upstream may emit before exiting zero.
"""
from __future__ import annotations

import json
import math

from harness._subproc import SubprocOutcome
from harness.base import Adapter, BuildCommand, HarnessError, ParsedOutput, RunSpec

_MODE_ENV = "GOOSE_MODE"
_BYPASS_MODE = "auto"


class GooseAdapter(Adapter):
    name = "goose"
    instructions_filename = ""  # Inline via --system=; no workdir projection.
    DEFAULT_MODEL = ""  # Leave provider/model selection to the caller's goose config.
    permission_bypass_args = ()  # Bypass is GOOSE_MODE=auto in the child env, not argv.
    config_home_env = "GOOSE_PATH_ROOT"

    def reported_model(self, spec: RunSpec) -> str | None:
        return spec.model or None

    def build_command(self, spec: RunSpec) -> BuildCommand:
        resolved = self.resolve_run_spec(spec)
        env: dict[str, str] = {}
        if spec.permission_policy == "bypass":
            explicit = spec.env.get(_MODE_ENV)
            if explicit is not None and explicit != _BYPASS_MODE:
                raise HarnessError(
                    f"permission_policy='bypass' conflicts with env[{_MODE_ENV!r}]={explicit!r}; "
                    f"bypass requires {_MODE_ENV}={_BYPASS_MODE}",
                    code="invalid-options",
                )
            env[_MODE_ENV] = _BYPASS_MODE
        args = ["run", "--quiet", "--output-format", "stream-json"]
        if resolved.model:
            args += ["--model", resolved.model]
        if spec.instructions is not None:
            args.append(f"--system={spec.instructions}")
        # Equals form keeps leading '-' prompts (and the empty prompt) as the value.
        args.append(f"--text={spec.prompt}")
        return self.finalize_command(spec, cmd="goose", args=args, env=env)

    def parse_output(self, spec: RunSpec, outcome: SubprocOutcome) -> ParsedOutput:
        tokens_in, tokens_out, cost, raw = _parse_goose_events(outcome.stdout)
        return {"cost_usd": cost, "tokens_in": tokens_in, "tokens_out": tokens_out, "raw": raw}


def _parse_goose_events(stdout: str) -> tuple[int | None, int | None, float | None, list[dict] | None]:
    events: list[dict] = []
    complete: dict | None = None
    for line in stdout.split("\n"):
        line = line.strip()
        if not line:
            continue
        try:
            # Strict JSON like the TS port: NaN/Infinity literals invalidate the line.
            event = json.loads(line, parse_constant=_reject_constant)
        except ValueError:
            continue
        if not isinstance(event, dict) or not isinstance(event.get("type"), str):
            continue
        events.append(event)
        if event["type"] == "complete":
            complete = event
    if not events:
        return None, None, None, None
    if complete is None:
        return None, None, None, events
    return (
        _token_count(complete.get("input_tokens")),
        _token_count(complete.get("output_tokens")),
        _cost(complete.get("cost_usd")),
        events,
    )


def _reject_constant(name: str) -> object:
    raise ValueError(f"non-standard JSON constant {name!r}")


def _token_count(value: object) -> int | None:
    if isinstance(value, bool):
        return None
    if isinstance(value, int):
        return value if value >= 0 else None
    if isinstance(value, float) and value >= 0 and math.isfinite(value) and value.is_integer():
        return int(value)
    return None


def _cost(value: object) -> float | None:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return None
    if value < 0 or not math.isfinite(value):
        return None
    return float(value)
